packages/darklang/stdlib/json.dark

```diff
@@ -0,0 +1,113 @@
+          | CantMatchWithType of
+            PACKAGE.Darklang.LanguageTools.RuntimeTypes.TypeReference *
+            String *
+            JsonPath.JsonPath
+          | EnumExtraField of rawJson: String * JsonPath.JsonPath
+          | EnumMissingField of
+            PACKAGE.Darklang.LanguageTools.RuntimeTypes.TypeReference *
+            index: Int *
+            JsonPath.JsonPath
+          | EnumInvalidCasename of
+            PACKAGE.Darklang.LanguageTools.RuntimeTypes.TypeReference *
+            caseName: String *
+            JsonPath.JsonPath
+          | EnumTooManyCases of
+            PACKAGE.Darklang.LanguageTools.RuntimeTypes.TypeReference *
+            cases: List<String> *
+            JsonPath.JsonPath
+          | RecordDuplicateField of fieldName: String * JsonPath.JsonPath
+          | RecordMissingField of fieldName: String * JsonPath.JsonPath
+          | NotJson
+
+        let toString (e: ParseError) : String =
+          match e with
+          | CantMatchWithType(typ, json, path) ->
+            "Can't parse JSON `"
+            ++ json
+            ++ "` as type `"
+            ++ PACKAGE.Darklang.PrettyPrinter.RuntimeTypes.typeReference typ
+            ++ "` at path: `"
+            ++ JsonPath.toString path
+            ++ "`"
+
+          // Can't parse JSON due to an extra argument `2.0` at path: `root.Enh[2]`
+          | EnumExtraField(rawJson, path) ->
+            "Can't parse JSON due to an extra argument `"
+            ++ rawJson
+            ++ "` at path: `"
+            ++ JsonPath.toString path
+            ++ "`"
+
+          // Can't parse JSON because argument 4 (`Int`) is missing at path: `root.Enh`
+          | EnumMissingField(typ, index, path) ->
+            "Can't parse JSON because argument "
+            ++ Int.toString index
+            ++ " (`"
+            ++ PACKAGE.Darklang.PrettyPrinter.RuntimeTypes.typeReference typ
+            ++ "`) is missing at path: `"
+            ++ JsonPath.toString path
+            ++ "`"
+
+          // Can't parse JSON as `Wrong` is not a known case of type `PrettyLikely` at path: `root`
+          | EnumInvalidCasename(typ, caseName, path) ->
+            "Can't parse JSON as `"
+            ++ caseName
+            ++ "` is not a known case of type `"
+            ++ PACKAGE.Darklang.PrettyPrinter.RuntimeTypes.typeReference typ
+            ++ "` at path: `"
+            ++ JsonPath.toString path
+            ++ "`"
+
+          // Can't parse JSON with multiple fields (`Enh` and `Extra`) as type `PrettyLikely` at path: `root`
+          | EnumTooManyCases(typ, cases, path) ->
+            let cases = cases |> List.map (fun c -> $"`{c}`") |> String.join ", "
+
+            $"Can't parse JSON with multiple fields ({cases}) as type `"
+            ++ PACKAGE.Darklang.PrettyPrinter.RuntimeTypes.typeReference typ
+            ++ "` at path: `"
+            ++ JsonPath.toString path
+            ++ "`"
```

Comment:
I don't think this error case is super clear without users knowing how we serialize Enums.
(i.e. "Can't parse JSON with multiple fields" in particular is missing some context)

"For the type `type Option<'x> = Some<'x> | None`, a value `Some 1` is serialized as `{ Some: [1] }` - if an object such as `{ Some: [1], Other: [] }` is presented, we're not sure how to parse it" may be a useful addendum?

Reply:
I think users will have to know how we serialize Enums or else they won't know what's wrong.

This error isn't a real error. Unlike an RTE or other compiler error, this is something the user can use to print the Error enum at runtime. So looking up types feels outside the scope here, it's really just a pretty print.

It would be nice to do better, but I can't see how with that constraint.